src/components/OIDCLanding.js

```diff
@@ -1,155 +1,185 @@
+// running multiple times, which is explictly called out as an anti-pattern
+// in that same document. What gives?!? Once we know OTP exchange was
+// successful, then we need to initialize the session, including populating
+// the redux, which means application init can't happen completely outside the
+// component lifecycle where the store isn't available.
+//
+const OTP_EXCHANGE_SUCCESS = '@folio/stripes/core::OTPExchangeSuccess';
+let otpError = null;
+
+export const exchangeOtp = (setToken, setTenant) => {
+  otpError = null;
+  const urlParams = new URLSearchParams(window.location.search);
+  const otp = urlParams.get('code');
+  const { tenant, clientId } = getLoginTenant(okapiConfig, config);
+
+  const redirectUri = getOIDCRedirectUri(tenant, clientId);
+
+  if (otp) {
+    fetch(`${okapiConfig.url}/authn/token?code=${otp}&redirect-uri=${redirectUri}`, {
+      credentials: 'include',
+      headers: { 'X-Okapi-tenant': tenant, 'Content-Type': 'application/json' },
+      mode: 'cors',
+    })
+      .then((resp) => {
+        if (resp.ok) {
+          return resp.json()
+            .then((json) => {
+              return setToken({
+                // if AT/RT values are missing in the response from _self,
+                // store near-future expiration values instead of the (missing)
+                // actual values. given that we arrived here, we know the request
+                // was successful, i.e. cookies are present, even though their
+                // values are (oddly) missing in the response body.
+                atExpires: json.accessTokenExpiration ? new Date(json.accessTokenExpiration).getTime() : Date.now() + (60 * 1000),
+                rtExpires: json.refreshTokenExpiration ? new Date(json.refreshTokenExpiration).getTime() : Date.now() + (2 * 60 * 1000),
+              });
+            })
+            .then(() => {
+              // for login itself we're storing selected tenant and clientId in the url
+              // but we still need to store tenantId in localStorage for logout purposes
+              // `logout` function in loginServices.js provides details about this.
+              setTenant(tenant);
+
+              // emit OPT_EXCHANGE_SUCCESS just in case the component mounted
+              // before OTP exchange completed. otherwise, the component would
+              // not know about the status of OTP exchange, and would never
+              // begin session-init.
+              window.dispatchEvent(new Event(OTP_EXCHANGE_SUCCESS));
+            });
+        } else {
+          return resp.json().then((error) => {
+            throw error;
+          });
+        }
+      })
+      .catch(e => {
+        // eslint-disable-next-line no-console
+        console.error('@@ Oh, snap, OTP exchange failed!', e);
+        // copy the error into the global otpError
+        otpError = e;
+      });
+  }
+};
+
+exchangeOtp(setTokenExpiry, storeLogoutTenant);
```

Comment:
@zburke what is the purpose of calling this function on module level? it's called in tests when at least smth is imported from stripes-core

Reply:
These are good points, @NikitaSedyx. This PR solves a real problem, but it does so in a pretty gross way. See PR #1656 to revert this PR. I reopened the ticket and will work on a better approach.